dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/export/FileResourceStream.java

```diff
@@ -43,5 +53,96 @@
+            throw new ConflictException(
+                "The content is being processed and is not available yet. Try again later.");
+          } catch (IOException e) {
+            throw new ConflictException(
+                "Failed fetching the file from storage",
+                "There was an exception when trying to fetch the file from the storage backend. "
+                    + "Depending on the provider the root cause could be network or file system related.");
+          }
+        });
+  }
+
+  /**
+   * Creates a {@code FileResourceStream} of an image of the given dimension. Dimension defaults to
+   * the original dimensions if the given {@code dimension} is {@code null}.
+   *
+   * <p>Images have to be fetched and buffered in memory for image dimension other than the
+   * original. This is due to us currently only storing the originals content length and md5 hash in
+   * the DB. We thus need to read the image and compute its content length and md5 hash for them to
+   * be available for cache validation.
+   *
+   * @param dimension the dimension of the image to create a stream for
+   * @throws BadRequestException when the file resource is not an image, does not support multiple
+   *     dimensions or does not have multiple dimension files stored
+   */
+  @Nonnull
+  public static FileResourceStream ofImage(
+      FileResourceService fileResourceService,
+      FileResource fileResource,
+      @CheckForNull ImageFileDimension dimension)
+      throws BadRequestException, ConflictException {
+    // The FileResource only stores the storageKey, contentLength and md5Hash of the original image.
+    // At least for now we are losing the benefit of not fetching the file from storage if the
+    // client already has an up-to-date version of the image in the given dimension other than the
+    // original. We have to fetch and compute the length and hash of the image again.
+    ImageFileDimension imageDimension =
+        ObjectUtils.firstNonNull(dimension, ImageFileDimension.ORIGINAL);
+    FileResourceStream fileResourceStream = new FileResourceStream(fileResource);
+    if (imageDimension != ImageFileDimension.ORIGINAL) {
+      byte[] content;
+      try {
+        content = fileResourceService.copyImageContent(fileResource, imageDimension);
+        fileResourceStream.inputStreamSupplier = () -> new ByteArrayInputStream(content);
+      } catch (NoSuchElementException e) {
+        // Note: we are assuming that the file resource is not available yet. The same approach
+        // is taken in other file endpoints or code relying on the storageStatus = PENDING.
+        // All we know for sure is the file resource is in the DB but not in the store.
+        throw new ConflictException(
+            "The content is being processed and is not available yet. Try again later.");
+      } catch (IOException e) {
+        throw new ConflictException(
```

Comment:
Is it also worth having unit tests for all these different ifs and exceptions that are thrown?

Reply:
These are tested via the controller tests 👌🏻